std/string.d

```diff
@@ -5798,36 +5798,57 @@ unittest
  * function, or any of the conversion functions.
  *
  * Params:
- *     s = the string to check
+ *     s = the string or random access range to check
  *     bAllowSep = accept separator characters or not
  *
  * Returns:
  *     $(D bool)
  */
-bool isNumeric(S)(S s, bool bAllowSep = false) if (isSomeString!S)
+bool isNumeric(S)(S s, bool bAllowSep = false) if (isSomeString!S ||
+    (isRandomAccessRange!S &&
+    hasSlicing!S &&
+    isSomeChar!(ElementEncodingType!S) &&
+    !isInfinite!S))
 {
     import std.algorithm.comparison : among;
+    import std.ascii : isASCII;
+    import std.utf : byChar;
 
-    if (s.empty)
+    // auto-decoding special case, we're only comparing to characters
+    // in the ASCII range so there's no reason to decode
+    static if (isSomeString!S)
+    {
+        import std.utf : byCodeUnit;
+        auto codeUnits = s.byCodeUnit;
+    }
+    else
+    {
+        alias codeUnits = s;
+    }
+
+    if (codeUnits.empty)
         return false;
 
     // Check for NaN (Not a Number) and for Infinity
-    if (s.among!((a, b) => icmp(a, b) == 0)
+    if (codeUnits.among!((a, b) => icmp(a, b.byChar) == 0)
             ("nan", "nani", "nan+nani", "inf", "-inf"))
         return true;
 
-    if (s.front == '-' || s.front == '+')
-        s.popFront;
+    if (codeUnits.front == '-' || codeUnits.front == '+')
+        codeUnits.popFront;
 
-    immutable iLen = s.length;
+    immutable iLen = codeUnits.length;
     bool bDecimalPoint, bExponent, bComplex, sawDigits;
 
-    for (size_t i = 0; i < iLen; i++)
+    for (int i = 0; i < iLen; ++i)
```

Comment:
What's the point of making `i` an `int`? This just makes it take fewer iterations to reach problems, as far as I see.